Add support for long Windows paths

mistafunk opened the pull request:
### Description of Change(s)

Prefix all file system paths longer than 248 characters with "\\\\?\\" before passing into Win32 File I/O API functions.

### Fixes Issue(s)

Avoids errors when reading/writing USD files on long Windows paths.

- [X] I have verified that all unit tests pass with the proposed changes
- [X] I have submitted a signed Contributor License Agreement

### Notes
* ~~The CLA is pending on my side.~~
* This is my first stab at fixing this problem and also my first time contributing larger changes to USD.
* ~~This PR is not complete yet (Arch and Tf test coverage). I submit it early to get some basic feedback.~~
* ~~This PR contains temporary/helper commits not intended for merge.~~

mistafunk committed:
Arch: long path handling for Win32 I/O API calls

mistafunk committed:
Arch: add tests for Windows long path handling

mistafunk committed:
git ignore CLion project files

mistafunk commented:
Looking into the compile issues...

meshula commented on pxr/base/arch/fileSystem.cpp:
```diff
@@ -1387,4 +1429,35 @@ void ArchFileAdvise(
+#if defined(ARCH_OS_WINDOWS)
+
+ARCH_API
+std::wstring ArchHandleLongWindowsPaths(const std::wstring& path)
```
minor note, the pattern 

`ArchHandleLongWindowsPaths(ArchWindowsUtf8ToUtf16(path));`

is very common in this patch. What about overloading this function instead, with a const char* path overload to match the callsites?

mistafunk replied:
Good idea, done.

meshula commented:
Looks promising! Out of curiosity, I noticed you removed the Debug target, which I assume is a temporary development detail. Is there some aspect of Debug that was getting in the way, or is it just a patch to streamline a workflow you have?

tallytalwar commented:
Filed as internal issue #USD-7967

mistafunk committed:
Arch: hide ArchHandleLongWindowsPaths

mistafunk committed:
Arch: additional test coverage

mistafunk committed:
Arch/Tf: introduce ArchTouchFile and forward from TfTouchFile

mistafunk committed:
Arch: add commonly used overload for _ArchHandleLongWindowsPaths

mistafunk commented:
> Looks promising! Out of curiosity, I noticed you removed the Debug target, which I assume is a temporary development detail. Is there some aspect of Debug that was getting in the way, or is it just a patch to streamline a workflow you have?

Thanks. I saw CMake install errors regarding `testArchStackTrace.pdb` in non-Release modes. I don't grok these yet, but it seems to be related to the VS generator. Switching to Ninja makes these go away.

Anyways, I'm working on another PR iteration and will trigger a review request once I'm ready.

erikaharrison-adsk commented:
Re: testArchStackTrace.pdb, see https://github.com/PixarAnimationStudios/USD/pull/2179

mistafunk committed:
Arch: don't leak temp file in ArchPWrite and ArchPRead test

mistafunk commented:
@meshula I've rebased the PR to latest dev and expanded the test coverage a bit.

Please note, I've yet to backport these changes to our product and test them there. But I consider the changes done.

mistafunk commented:
Before merging, the  `_ArchHandleLongWindowsPaths` also needs to remove dot and dotdot in the paths as well as two (or more) consecutive backslashes.

(Stumbled over this while adding long path support to another library.)

johnk-realtime commented:
Hey guys, just checking on the status of this as it appears the discussion has been inactive for a while. Besides the review is there anything else that's preventing this from moving forwards?

We are experiencing this issue at the moment and considering our options :)

mistafunk commented:
Hi, the ball is with Pixar :-) I'm happy to receive feedback and work some more on this...

spiffmon commented:
@mistafunk , it's taken awhile, but this has, coincidentally, recently finally risen to near top of queue, and we'll be having a look shortly.  Can't promise it will make it in to 24.08, but soon.

gitamohr commented on pxr/base/arch/fileSystem.cpp:
```diff
@@ -648,21 +762,27 @@ MakeUnique(
 
+namespace {
+
+const std::string TMPDIR_FMT = std::string("%s") + ARCH_PATH_SEP + "%s.XXXXXX";
```
Can this just be a literal?  `"%s" ARCH_PATH_SEP "%s.XXXXXX"` ?

mistafunk replied:
good catch, will turn this into a `constexpr const char*`

mistafunk replied:
done

jesschimein commented:
/AzurePipelines run

mistafunk commented:
/AzurePipelines run

Commit:
Arch: move ArchNormPath helpers to top, so we can use it in Windows long path handling

Commit:
Arch: templatize _NormPath to support wide strings

Commit:
Arch: normalize the long path to satisfy Win32 long path requirements (e.g. no dotdot and dot)

mistafunk commented:
@spiffmon @meshula I've rebased the branch onto latest dev and addressed the one remaining comment as well as made sure the long paths are normalized (removal of dotdot and dot) as per MS specs.

jesschimein commented:
/AzurePipelines run

mistafunk commented:
Oops, let me take care of the compile issues on Linux/macOS...

mistafunk commented:
Force pushed the last two commits and fixed Linux build by adding a missing `typename` (hopefully this takes care of macOS as well... otherwise I'll fire up my macbook in the next few days)

mistafunk commented:
@spiffmon @meshula resolved merge conflicts and rebased on `dev`... soon this PR turns 2 years old, do you think we manage to merge this before it turns into a teenager? :-)

jesschimein commented:
/AzurePipelines run

demiaster commented:
Hello!
We are also interested in seeing this feature through.
Is there any outstanding issues that need resolving?

Thank you!

spiffmon commented:
Apologies, this has languished on our work list for awhile.  We've scheduled it for an upcoming release this year, and given the time that has elapsed since @mistafunk made the changes we requested, we will handle any further updating/changes we think necessary.
